### PR TITLE
Fixing the index of the byte used for setting & retrieveing the proje…

### PR DESCRIPTION
**Pull request by ahshahid**
…ction bit set. Fixes SNAP-2890

Correcting the position of the byte used to set the projection bit in ColumnFormatIterator
As the index is 0 based, using column number -1, to identify the position of the byte &  the bit for masking.
The issue is that say if the max column in projection is 128, the number of bytes needed to store bits is 16 ( 0 to 15). But for 128th column, the current code gets index as 16 ( instead of 15)
It is possible that the fix may be needed at other places too?

Added bug test

### CODE REVIEW

**Comment by ahshahid** [comment]
@sumwale @kneeraj @suranjan @ashetkar , can you pls review this asap, as the customer project is stuck on this issue.

**Comment by ahshahid** [comment]
@sumwale @kneeraj @suranjan @ashetkar , pls review...

File: cluster/src/test/scala/org/apache/spark/sql/store/BugTest.scala [hunk 1]
```diff
@@ -19,10 +19,15 @@ package org.apache.spark.sql.store
 import java.io.{BufferedReader, FileReader}
 import java.sql.{DriverManager, SQLException}
 
+import scala.tools.nsc.interpreter
+import scala.tools.nsc.interpreter.session
+
```
Comment (sumwale) [inline cluster/src/test/scala/org/apache/spark/sql/store/BugTest.scala]: these imports are required?

File: cluster/src/test/scala/org/apache/spark/sql/store/BugTest.scala [hunk 2]
```diff
@@ -475,5 +480,76 @@ class BugTest extends SnappyFunSuite with BeforeAndAfterAll {
...
+      "using column ")
+
+    val params = Array.fill(numCols)('?').mkString(",")
+    val insertStr = s"insert into test1 values (${params})"
+    val ps = conn.prepareStatement(insertStr)
+    for (i <- 0 until 1000) {
+      for (j <- 1 until numCols + 1) {
+        ps.setString(j, j.toString)
+      }
+      ps.executeUpdate()
```
Comment (sumwale) [inline cluster/src/test/scala/org/apache/spark/sql/store/BugTest.scala]: better to do batch insert which will be much more efficient and reduce test running time: ps.addBatch(), and ps.executeBatch() outside the loop

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatIterator.scala [hunk 3]
```diff
@@ -75,7 +75,7 @@ final class ColumnFormatIterator(baseRegion: LocalRegion, projection: Array[Int]
       val maxProjection = Ints.max(projection: _*)
       val bitset = new Array[Long](UnsafeRow.calculateBitSetWidthInBytes(maxProjection) >>> 3)
       for (p <- projection) {
-        BitSet.set(bitset, Platform.LONG_ARRAY_OFFSET, p)
+        BitSet.set(bitset, Platform.LONG_ARRAY_OFFSET, p-1)
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatIterator.scala]: missing spaces: p - 1

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatIterator.scala [hunk 4]
```diff
@@ -217,7 +217,7 @@ final class ColumnFormatIterator(baseRegion: LocalRegion, projection: Array[Int]
           val tableColumn = ColumnDelta.tableColumnIndex(columnIndex)
           tableColumn > 0 &&
               BitSet.isSet(projectionBitSet, Platform.LONG_ARRAY_OFFSET,
-                tableColumn, projectionBitSet.length)
+                tableColumn -1 , projectionBitSet.length)
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatIterator.scala]: missing space after -

File: cluster/src/test/scala/org/apache/spark/sql/store/BugTest.scala (hunk 1 above)
Comment (ahshahid) [reply cluster/src/test/scala/org/apache/spark/sql/store/BugTest.scala]: will remove it..

**Pull request merged**